### PR TITLE
Devs/hlod sub tree on quad tree

### PR DESCRIPTION
**Pull request by JangkyuSeo**
### Purpose of this PR
Added new UI for choosing SpaceSplitter with option values.
image

Also a new option is added, Use sub HLOD tree. It can split HLOD tree to several sub HLOD tree.
image

Each Controller is created for each SubTree, and a new GameObject is created and put it there.
image

---
### Release Notes
Added new UI for choosing SpaceSplitter with option values.

---
### Testing status
Manual test
All unit test are passed
Added new unit test for HLOD sub trees.

---
### Overall Product Risks
**Technical Risk**: Medium

**Halo Effect**: None, Medium

---

### CODE REVIEW

File: com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs [hunk 1]
```diff
@@ -135,9 +247,73 @@ public SpaceNode CreateSpaceTree(Bounds initBounds, float chunkSize, Transform t
...
+            for (int i = 0; i < targetBoundList.Count; ++i)
+            {
+                targetObjectsList.Add(new List<TargetInfo>());
+            }
+
+            foreach (var targetInfo in targetInfoList)
+            {
+                for (int bi = 0; bi < targetBoundList.Count; ++bi)
+                {
+                    if (targetBoundList[bi].Contains(targetInfo.Bounds.center))
```
[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs:289]: The object must be included to bounds and must be included in one.. The center of bounds is center position of objects. Also, targetBoundList is not overlapped each other. Therefore, If check that center is included, it is always included in one bounds.

File: com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs [hunk 2]
```diff
@@ -135,9 +247,73 @@ public SpaceNode CreateSpaceTree(Bounds initBounds, float chunkSize, Transform t
...
+        {
+            if (m_autoCaclLooseSize == true)
+            {
+                //If the chunk size is small, there is a problem that it may get caught in an infinite loop.
+                //So, the size can be determined according to the chunk size.
+                return Mathf.Min(chunkSize * 0.3f, 5.0f);
+            }
+            else
+            {
+                return m_looseSizeFromOptions;
```
[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs:310]: I realized what this logic mean ><
I totally wrong.
It fixed loose size to 5.0 and limit 30% of chunk size.
That means,  we don't have auto calculation ><
I will change it.

File: com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs [hunk 3]
```diff
@@ -65,5 +68,10 @@ public void OnAfterDeserialize()
                 m_idTable[m_idList[i]] = i;
             }
         }
+
+        private int EncodeID(int controllerID, int id)
+        {
+            return (controllerID & 0xff) << 24 | id;
```
[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs:74]: You're right. 
I want to make UI guide somewhere but I couldn't find a place where right place. 
So I will provide guidline by the manual.

[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs:74]: Also, the limitation is per HLOD component.
So 256 controllers per one HLOD, I think that is enough.

[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs:74]: I added message box when there are more than 256 SubTrees.

File: com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (katana7755) [reply com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs:289]: Oh yes!!! I missed that point!!! Great to know :)

File: com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (katana7755) [reply com.unity.hlod/Runtime/Serializer/UserDataSerializerBase.cs:74]: Yup! I saw your new commit. It looks perfect to me!

File: com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (katana7755) [reply com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs:310]: >< great finding!!!

[the remark this answers is not in the document]
Comment (JangkyuSeo) [reply com.unity.hlod/Editor/SpaceManager/QuadTreeSpaceSplitter.cs]: I removed auto calculate UI.

**Review by katana7755 — APPROVED** [review]
All look good to me! Thanks for inviting me as a reviewer for the PR!

**Pull request merged**